### PR TITLE
Support per host configuration

### PR DESCRIPTION
**Pull request by lexmag**

### CODE REVIEW

File: lib/dicon/secure_shell.ex [hunk 1]
```diff
@@ -38,7 +40,9 @@ defmodule Dicon.SecureShell do
     result =
       with :ok <- ensure_started(),
            {:ok, conn} <- :ssh.connect(host, port, opts, @timeout),
-        do: {:ok, conn}
+           {:ok, sftp_channel} <- :ssh_sftp.start_channel(conn, [timeout: @timeout]) do
```
Comment (whatyouhide) [inline lib/dicon/secure_shell.ex]: Do we need brackets here for `[timeout: @timeout]`?

Comment (lexmag) [inline lib/dicon/secure_shell.ex]: We don't. :)

Comment (whatyouhide) [inline lib/dicon/secure_shell.ex]: Phrased badly, sorry 😄 Do we _want_ brackets here? :)

File: lib/mix/tasks/dicon.deploy.ex [hunk 2]
```diff
@@ -60,4 +62,16 @@ defmodule Mix.Tasks.Dicon.Deploy do
...
   end
+
+  defp write_custom_config(conn, name, target_dir, version) do
+    if config = host_config(name) do
+      content = :io_lib.format('~p.\n', [config])
+      target_sub_dir = ["/releases/", version, ?/]
+      config_sub_path = [target_sub_dir, "custom.config"]
+      Executor.write_file(conn, [target_dir, config_sub_path], content)
+      vm_args_path = [target_dir, target_sub_dir, "vm.args"]
+      content = ["-config ", [".", config_sub_path]]
```
Comment (whatyouhide) [inline lib/mix/tasks/dicon.deploy.ex]: As far as I can tell, we don't end the `vm.args` file with a newline here. Is this indented behaviour?

Comment (lexmag) [inline lib/mix/tasks/dicon.deploy.ex]: Oops, didn't push that commit.

File: lib/mix/tasks/dicon.deploy.ex [hunk 3]
```diff
@@ -60,4 +62,16 @@ defmodule Mix.Tasks.Dicon.Deploy do
...
   end
+
+  defp write_custom_config(conn, name, target_dir, version) do
+    if config = host_config(name) do
+      content = :io_lib.format('~p.\n', [config])
+      target_sub_dir = ["/releases/", version, ?/]
+      config_sub_path = [target_sub_dir, "custom.config"]
+      Executor.write_file(conn, [target_dir, config_sub_path], content)
+      vm_args_path = [target_dir, target_sub_dir, "vm.args"]
+      content = ["-config ", [".", config_sub_path], ?\n]
```
Comment (whatyouhide) [inline lib/mix/tasks/dicon.deploy.ex]: Why the sublist here for `[".", config_sub_path]`? It could be `["-config .", config_sub_path, ?\n]` maybe or am I missing something?

File: test/dicon/executor_test.exs [hunk 4]
```diff
@@ -12,6 +12,9 @@ defmodule Dicon.ExecutorTest do
     def exec(_conn, :fail, _device), do: {:error, "exec failed"}
     def exec(_conn, _command, _device), do: :ok
 
+    def write_file(_conn, :fail, "fail", _mode), do: {:error, "copy failed"}
```
Comment (whatyouhide) [inline test/dicon/executor_test.exs]: The error message here and in the test may be better as `"write failed"`, wdyt?

Comment (lexmag) [inline test/dicon/executor_test.exs]: Yeah, it should be so.

File: test/dicon/executor_test.exs [hunk 5]
```diff
@@ -45,4 +48,15 @@ defmodule Dicon.ExecutorTest do
     message = "(in Dicon.ExecutorTest.FakeExecutor) copy failed"
     assert_raise Mix.Error, message, fn -> Executor.copy(conn, :fail, :fail) end
   end
+
+  test "write_file/4" do
+    conn = Executor.connect(:whatever)
+
+    assert Executor.write_file(conn, :target, "content") == :ok
```
Comment (whatyouhide) [inline test/dicon/executor_test.exs]: The spec for `write_file/4` says the target should be `char_data`, maybe we should respect that here and use something like `write_file(conn, "./target", "content")` to show it's a path even?

Comment (lexmag) [inline test/dicon/executor_test.exs]: It'll be a separate commit for everything in `executor_test.exs`.

File: lib/mix/tasks/dicon.deploy.ex (hunk 3 above)
Comment (lexmag) [inline lib/mix/tasks/dicon.deploy.ex]: Just to show a config path.

Comment (whatyouhide) [inline lib/mix/tasks/dicon.deploy.ex]: Yep, I suspected that, but maybe it raises more questions than answers? 😃 No strong preferences anyways!

File: lib/dicon/executor.ex [hunk 6]
```diff
@@ -46,7 +50,7 @@ defmodule Dicon.Executor do
   def connect(authority) do
     executor = Application.get_env(:dicon, :executor, SecureShell)
     case executor.connect(authority) do
-      {:ok, id}        -> %__MODULE__{executor: executor, id: id}
+      {:ok, conn} -> %__MODULE__{executor: executor, conn: conn}
```
Comment (whatyouhide) [inline lib/dicon/executor.ex:53]: :suspect: don't think I didn't see this whitespace murder 😈

File: lib/mix/tasks/dicon.deploy.ex [hunk 7]
```diff
@@ -60,4 +62,16 @@ defmodule Mix.Tasks.Dicon.Deploy do
...
+  defp write_custom_config(conn, name, target_dir, version) do
+    if config = host_config(name) do
+      content = :io_lib.format('~p.\n', [config])
+      target_sub_dir = ["/releases/", version, ?/]
+      config_sub_path = [target_sub_dir, "custom.config"]
+      Executor.write_file(conn, [target_dir, config_sub_path], content)
+      vm_args_path = [target_dir, target_sub_dir, "vm.args"]
+      content = ["-config ", ?., config_sub_path, ?\n]
+      Executor.write_file(conn, vm_args_path, content, :append)
+    end
```
Comment (whatyouhide) [inline lib/mix/tasks/dicon.deploy.ex:77]: This is a subjective opinion ⚠️: I find this function slightly hard to follow because of the "wall-of-text" effect, so I'll suggest some layout edits that may make it more readable (they do IMO).

``` elixir
defp write_custom_config(conn, name, target_dir, version) do
  if config = host_config(name) do
    # First, we compute the stuff we need for both writing custom.config and appending to vm.args
    target_sub_dir = ["/releases/", version, ?/]
    config_sub_path = [target_sub_dir, "custom.config"]

    # Then, after a good old newline, we write custom.config
    content = :io_lib.format('~p.\n', [config])
    Executor.write_file(conn, [target_dir, config_sub_path], content)

    # and in a "separate" paragraph, we write vm.args
    vm_args_path = [target_dir, target_sub_dir, "vm.args"]
    content = ["-config ", ?., config_sub_path, ?\n]
    Executor.write_file(conn, vm_args_path, content, :append)
  end
end
```

Wdyt? This way the function seems more divided into logical "paragraphs", something I find really really useful and easier to read :)

File: lib/dicon/executor.ex (hunk 6 above)
Comment (lexmag) [inline lib/dicon/executor.ex:53]: 👹
